src/withdrawal_limit/component.cairo

```diff
@@ -0,0 +1,126 @@
+            if remaining_quota == 0 {
+                return self.get_daily_withdrawal_limit(:token);
+            }
+            remaining_quota - constants::REMAINING_QUOTA_OFFSET
+        }
+    }
+
+    #[generate_trait]
+    pub impl InternalImpl<
+        TContractState, +HasComponent<TContractState>, +ITokenBridge<TContractState>
+    > of InternalTrait<TContractState> {
+        // Sets the remaining withdrawal quota for today.
+        fn set_remaining_withdrawal_quota(
+            ref self: ComponentState<TContractState>, token: ContractAddress, amount: u256
+        ) {
+            let now = get_block_timestamp();
+            let day = now / constants::SECONDS_IN_DAY;
+            self
+                .remaining_intraday_withdraw_quota
+                .write((token, day), amount + constants::REMAINING_QUOTA_OFFSET);
+        }
+
+        // Returns the remaining withdrawal quota for today.
+        fn read_withdrawal_quota_slot(
+            self: @ComponentState<TContractState>, token: ContractAddress
+        ) -> u256 {
+            let now = get_block_timestamp();
+            let day = now / constants::SECONDS_IN_DAY;
+            self.remaining_intraday_withdraw_quota.read((token, day))
+        }
+
+
+        // Try to withdraw an amount and if it succeeds, update the remaining withdrawal quota.
+        fn consume_withdrawal_quota(
+            ref self: ComponentState<TContractState>,
+            token: ContractAddress,
+            amount_to_withdraw: u256
+        ) {
+            let remaining_withdrawal_quota = self.get_remaining_withdrawal_quota(token);
+            // This function should be called only after checking that `is_withdrawal_limit_applied`
+            // is true. When limit withdrawal is disabled, the `remaining_withdrawal_quota` is
+            // BoundedInt::max(). We rely on that to limit the access only to cases where limit
+            // withdrawal is enabled.
+            assert(
+                remaining_withdrawal_quota < BoundedInt::max(), 'withdrawal_limit_applied ERROR'
+            );
+            assert(remaining_withdrawal_quota >= amount_to_withdraw, 'LIMIT_EXCEEDED');
+            self
+                .set_remaining_withdrawal_quota(
+                    :token, amount: remaining_withdrawal_quota - amount_to_withdraw
+                )
+        }
+
+        // Returns the full quota of the daily withdrawal limit for a given token.
+        // The calculation is based on the limit percentage and current token total supply.
+        // Note - while techincally, we're exposed to overflow error here, we consider that
+        // risk non-existant for any token of even the smallest value.
+        fn get_daily_withdrawal_limit(
+            self: @ComponentState<TContractState>, token: ContractAddress
+        ) -> u256 {
+            let total_supply = IERC20Dispatcher { contract_address: token }.total_supply();
```

Comment:
why do we use total_supply instead of balance of the contract

Reply:
I think the rationale is to have withdraw limit pct in total_supply 
[Reference](https://github.com/starknet-io/starkgate-contracts/blob/cairo-1/src/cairo/token_bridge.cairo#L331)